Toolchain selection

matthewbastien opened the pull request:
Fixes #756

plemarquand commented on src/toolchain/toolchain.ts:
```diff
@@ -181,12 +182,84 @@ export class SwiftToolchain {
      * @returns Folders for each Xcode install
      */
     public static async getXcodeInstalls(): Promise<string[]> {
+        if (process.platform !== "darwin") {
```
on line 150 the method `getXcodeDeveloperDir` calls `xcode-select -p`. If I choose another Xcode as my toolchain from within VSCode this still points to my system default toolchain.

plemarquand commented on src/toolchain/toolchain.ts:
```diff
@@ -181,12 +182,84 @@ export class SwiftToolchain {
+     *
+     * @returns an array of toolchain paths
+     */
+    public static getToolchainInstalls(): Promise<string[]> {
```
Is this Darwin only?

matthewbastien replied:
Yup. The other platforms don't have common directories where toolchains get installed. I've added a short circuit at the beginning of this function to avoid checking on non-darwin platforms.

plemarquand commented on src/toolchain/toolchain.ts:
```diff
@@ -181,12 +182,84 @@ export class SwiftToolchain {
+        } catch {
+            // Assume that there are no installations here
+        }
+        return [];
```
nit: move the return to the `catch` to make it more clear that `[]` is returned for the error condition

matthewbastien replied:
done.

plemarquand commented on src/ui/ToolchainSelection.ts:
```diff
@@ -0,0 +1,198 @@
+            };
+        });
+    const toolchains = (await SwiftToolchain.getToolchainInstalls())
+        .sort((a, b) => (a > b ? -1 : 1)) // Reverse order
```
```suggestion
        .reverse()
```

matthewbastien replied:
done

plemarquand commented on src/ui/ToolchainSelection.ts:
```diff
@@ -0,0 +1,198 @@
+            swiftFolderPath: path.join(toolchainPath, "usr", "bin"),
+        }));
+    const swiftlyToolchains = (await SwiftToolchain.getSwiftlyToolchainInstalls())
+        .sort((a, b) => (a > b ? -1 : 1)) // Reverse order
```
```suggestion
        .reverse()
```

matthewbastien replied:
done

plemarquand commented on src/ui/ToolchainSelection.ts:
```diff
@@ -0,0 +1,198 @@
+    const selected = await vscode.window.showQuickPick<SelectToolchainItem>(
+        getQuickPickItems(ctx),
+        {
+            title: "Select the swift toolchain",
+            placeHolder: "Pick a swift toolchain",
```
```suggestion
            title: "Select the Swift toolchain",
            placeHolder: "Pick a Swift toolchain",
```

matthewbastien replied:
done

plemarquand commented on src/ui/ToolchainSelection.ts:
```diff
@@ -0,0 +1,198 @@
+        canSelectFiles: false,
+        canSelectFolders: true,
+        canSelectMany: false,
+        title: "Select the folder containing swift binaries",
```
```suggestion
        title: "Select the folder containing Swift binaries",
```

plemarquand commented on src/sourcekit-lsp/LanguageClientManager.ts:
```diff
@@ -113,9 +113,9 @@ export class LanguageClientManager {
-        this.singleServerSupport = workspaceContext.swiftVersion.isGreaterThanOrEqual(
-            new Version(5, 7, 0)
-        );
+        this.singleServerSupport = workspaceContext.toolchain
+            ? workspaceContext.toolchain.swiftVersion.isGreaterThanOrEqual(new Version(5, 7, 0))
+            : false;
```
```suggestion
        this.singleServerSupport = workspaceContext.toolchain?.swiftVersion.isGreaterThanOrEqual(new Version(5, 7, 0)) ?? false;
```

plemarquand commented on src/toolchain/toolchain.ts:
```diff
@@ -181,12 +182,84 @@ export class SwiftToolchain {
+            );
+            const swiftlyConfig: unknown = JSON.parse(swiftlyConfigRaw);
+            if (
+                swiftlyConfig === undefined ||
+                swiftlyConfig === null ||
+                typeof swiftlyConfig !== "object" ||
+                !("installedToolchains" in swiftlyConfig)
```
```suggestion
                !(swiftlyConfig instanceof Object) ||
                !("installedToolchains" in swiftlyConfig)
```

adam-fowler commented:
Haven't looked at code in any detail as I'm doing this on my phone. I assume you have included a valid state where there is no toolchain. 

Instead of checking do we have a toolchain everywhere why don't you create two stages to the extension initialisation one for the toolchain and one for the rest. At initialisation if you have no toolchain you are forced through the select toolchain UI and the rest of the extension doesn't initialize until a toolchain is either downloaded or selected.

matthewbastien commented:
> Haven't looked at code in any detail as I'm doing this on my phone. I assume you have included a valid state where there is no toolchain.
> 
> Instead of checking do we have a toolchain everywhere why don't you create two stages to the extension initialisation one for the toolchain and one for the rest. At initialisation if you have no toolchain you are forced through the select toolchain UI and the rest of the extension doesn't initialize until a toolchain is either downloaded or selected.

Fair point. I've made it so that the WorkspaceContext doesn't even get created if there's no available toolchain. This allows us to keep relatively the same behaviour as before this patch. However, the command registration now takes an optional WorkspaceContext and will display the toolchain error message if commands are invoked without a WorkspaceContext. I've also moved some of the functionality that doesn't require a toolchain outside of the WorkspaceContext (e.g. the reload extensions notification).

adam-fowler commented:
So far tested this on macOS (Although I didn't test with multiple Xcode's installed) and works quite well. One minor issue, if you have a local toolchain and then select a new toolchain but set it globally you are still using the original local toolchain. I think you should probably clear the local setting here, or add another dialog to ask do you want to clear it.

I'll do a code review later, but I did see a lot of the code is still dealing with possibility of an undefined toolchain, folder context etc, which should be unnecessary now given your description of the changes you made.

adam-fowler commented on src/commands.ts:
```diff
@@ -893,6 +939,10 @@ export function register(ctx: WorkspaceContext) {
         vscode.commands.registerCommand("swift.selectXcodeDeveloperDir", () =>
```
I guess this and the referenced function can be removed now, as select toolchain does the equivalent and more

matthewbastien replied:
Actually, I believe the DEVELOPER_DIR setting is still useful. There are cases where you would want to set it even though you've selected a public toolchain, for example.

adam-fowler replied:
Possibly but I was using it as a method to select the toolchain from an Xcode install. I think including this and `select toolchain...` is confusing to the user.

matthewbastien replied:
Fair enough. I've removed it, but a sufficiently advanced user can still use the configuration to set DEVELOPER_DIR if they need it.

adam-fowler commented on src/extension.ts:
```diff
@@ -42,11 +46,57 @@ export interface Api {
+            : undefined;
+        context.subscriptions.push(...commands.register(workspaceContext));
+
+        if (!toolchain && vscode.workspace.workspaceFolders) {
```
Why `&& vscode.workspace.workspaceFolders` someone can create a single swift file and still expect swift support.

adam-fowler reviewed (changes requested):
Here is an initial review of code. There is a load of code related to having an optional WorkspaceContext that can be reverted. 

We need to make sure we get the UI flow correct for toolchain installation if no toolchain is currently installed.

matthewbastien commented on src/ui/ToolchainSelection.ts:
```diff
@@ -0,0 +1,200 @@
+ * Open the installation page on Swift.org
+ */
+export async function downloadToolchain() {
+    if (await vscode.env.openExternal(vscode.Uri.parse("https://www.swift.org/install/"))) {
```
I'm thinking this will be used more by developers who are new to Swift. The install page is a little more friendly to that, imo.

matthewbastien commented on src/ui/ToolchainSelection.ts:
```diff
@@ -0,0 +1,200 @@
+ */
+export async function downloadToolchain() {
+    if (await vscode.env.openExternal(vscode.Uri.parse("https://www.swift.org/install/"))) {
+        const selected = await showReloadExtensionNotification(
```
We can't auto-detect when a new toolchain is installed unless we do some polling. The user needs to be notified that they need to reload once they're done installing.

adam-fowler replied:
Yeah it makes the flow a little more awkward, maybe the messages should be different for when someone selects download

matthewbastien commented on src/WorkspaceContext.ts:
```diff
@@ -37,23 +36,22 @@ import { TestCoverageReportProvider } from "./coverage/TestCoverageReport";
-    public currentDocument: vscode.Uri | null;
-    public outputChannel: SwiftOutputChannel;
+    public currentFolder: FolderContext | undefined;
+    public currentDocument: vscode.Uri | undefined;
```
I can't even remember why I changed it tbh... I can revert if you want.

adam-fowler replied:
yes please

matthewbastien commented on src/configuration.ts:
```diff
@@ -159,6 +159,58 @@ const configuration = {
     get path(): string {
         return vscode.workspace.getConfiguration("swift").get<string>("path", "");
     },
+    async setPath(
```
Done.

matthewbastien commented on src/configuration.ts:
```diff
@@ -159,6 +159,58 @@ const configuration = {
+                });
+                if (vscode.workspace.workspaceFolders.length > 1) {
+                    items.push({
+                        label: "workspace folders",
```
I'm leaning towards keeping this lower case. Microsoft seems to prefer lower case labels for their separators.

matthewbastien replied:
My latest patch removes this code.

matthewbastien commented on src/extension.ts:
```diff
@@ -42,11 +46,57 @@ export interface Api {
+        const workspaceContext = toolchain
+            ? await WorkspaceContext.create(outputChannel, toolchain)
+            : undefined;
+        context.subscriptions.push(...commands.register(workspaceContext));
```
The problem is that the commands will still appear in the command palette and running them yields an awful error message that just says something along the lines of "could not find the command". The way I have it set up now is that running the commands with an invalid toolchain pops up the toolchain error notification which they can use to resolve the issue.

adam-fowler replied:
The alternative is to use the `isActivated` context key to make the commands only available once the extension is fully activated with toolchain.

matthewbastien commented on src/extension.ts:
```diff
@@ -42,11 +46,57 @@ export interface Api {
+            .catch(error => {
+                outputChannel.log("Failed to discover Swift toolchain");
+                outputChannel.log(error);
+                contextKeys.createNewProjectAvailable = true;
```
See comment below about only registering certain commands. I'd like to leave the option for them to run the command and get a nice error message.

adam-fowler commented on src/WorkspaceContext.ts:
```diff
@@ -414,14 +380,17 @@ export class WorkspaceContext implements vscode.Disposable {
             return this.folders[index];
         }
         const folderContext = await FolderContext.create(folder, workspaceFolder, this);
+        if (!folderContext) {
```
This should be necessary anymore

adam-fowler commented on src/WorkspaceContext.ts:
```diff
@@ -476,8 +445,8 @@ export class WorkspaceContext implements vscode.Disposable {
-        // check we are using CodeLLDB
-        if (DebugAdapter.adapterName !== "lldb") {
+        // check we are using CodeLLDB and have a valid toolchain
+        if (!this.toolchain || DebugAdapter.adapterName !== "lldb") {
```
Don't need to check toolchain anymore

adam-fowler commented on src/toolchain/toolchain.ts:
```diff
@@ -242,12 +314,37 @@ export class SwiftToolchain {
+    public async getLLDB(): Promise<string> {
+        let lldbPath = path.join(this.swiftFolderPath, "lldb");
+        if (!(await pathExists(lldbPath))) {
+            if (process.platform !== "darwin") {
```
In what situation were you not finding LLDB?

adam-fowler replied:
ok

adam-fowler reviewed (commented):
A few more comments, nothing major.

matthewbastien commented on src/ui/ToolchainSelection.ts:
```diff
@@ -0,0 +1,200 @@
+        ...(swiftlyToolchains.length > 0
+            ? [new SeparatorItem("swiftly"), ...swiftlyToolchains]
+            : []),
+        new SeparatorItem("actions"),
```
I've added an action that will bring the user to the Swiftly install page.

matthewbastien commented on src/extension.ts:
```diff
@@ -42,11 +46,57 @@ export interface Api {
+            })
+        );
+
+        if (!workspaceContext) {
```
done

adam-fowler commented on src/extension.ts:
```diff
@@ -42,11 +46,53 @@ export interface Api {
+                // on toolchain config change, reload window
+                if (event.affectsConfiguration("swift.path")) {
+                    showReloadExtensionNotification(
+                        "Changing the Swift path requires the project be reloaded."
```
You may not have a project loaded, you may have multiple projects loaded...
```suggestion
                        "Changing the Swift path requires Visual Studio Code be reloaded."
```

matthewbastien replied:
done

adam-fowler commented on src/ui/ToolchainSelection.ts:
```diff
@@ -0,0 +1,200 @@
+    ];
+}
+
+/**
```
Ok yeah I mixed those two up. Yeah the two function names are very close `selectToolchain` could become `showSelectToolchainQuickPick` so something similar

matthewbastien replied:
done

adam-fowler reviewed (commented):
Two minor nits, apart from that I think we are good

matthewbastien commented:
> So far tested this on macOS (Although I didn't test with multiple Xcode's installed) and works quite well. One minor issue, if you have a local toolchain and then select a new toolchain but set it globally you are still using the original local toolchain. I think you should probably clear the local setting here, or add another dialog to ask do you want to clear it.
> 
> I'll do a code review later, but I did see a lot of the code is still dealing with possibility of an undefined toolchain, folder context etc, which should be unnecessary now given your description of the changes you made.

So, I did some more investigation into this. It appears that settings for individual workspace folders are actually ignored for `swift.path`. I've removed the option to configure workspace folder settings and added a warning about workspace settings taking precedence over user settings.

@adam-fowler could you take one more look at the code and see if we're good?

adam-fowler reviewed (approved):
This all looks good

I have thought of one issue though, perhaps we can resolve this in another PR

If I select an Xcode install that isn't the Xcode install returned by `xcode-select -p` (default). The calls to `SwiftToolchain.getLLDB` and `SwiftToolchain.getSDKPath` return the wrong paths ie from the default install. 

Should we set DEVELOPER_DIR environment var if we are setting the path to an Xcode install to ensure we get the correct LLDB and SDK path
